### PR TITLE
Fix Gridstore flushing, correctly implement drain persist

### PR DESCRIPTION
**Pull request by timvisee**
Fix a critical Gridstore bug that breaks flushing with alternating puts and deletes.

Gridstore uses deferred flushing. It means that we copy the list of pending updates when the flusher is created. Once flushing is invoked and updates are persisted, we remove these from the current list of updates (which might already have new entries). This 'removing' we call 'drain persisted'. It is critical to persist every change only once.

This drain function was broken and failed on some edge cases. For the problematic edge case we found in testing, I've added a new test.

The problem only appears when deferring a flush (for some time). If flushing immediately, the pending and persisted set of updates are equal and everything works as expected.

Bear in mind that I use the concept of a 'set' and 'unset' list in this PR. I plan to open a separate PR to apply this concept everywhere as it's much more easy to understand. Since it's only partially applied in this PR it might look like boilerplate.

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: lib/gridstore/src/tracker.rs [hunk 1]
```diff
@@ -649,4 +708,43 @@ mod tests {
             assert!(updates.drain_persisted_and_drop(&persisted));
         }
     }
+
+    #[test]
+    fn test_value_pointer_drain_set_unset_set() {
```
Comment (timvisee) [inline lib/gridstore/src/tracker.rs]: New test is here. This failed before fixing `drain_persisted_and_drop`.

File: lib/gridstore/src/tracker.rs [hunk 2]
```diff
@@ -124,20 +124,69 @@ impl PointerUpdates {
...
 
-        // Remove all persisted pointers from history
-        self.history
-            .retain(|pointer| !persisted_pointers.contains(pointer));
+        // In current history, range of unset entries
+        let unset_range = if self.latest_is_set {
+            0..self.history.len().saturating_sub(1)
+        } else {
+            0..self.history.len()
+        };
```
Comment (timvisee) [inline lib/gridstore/src/tracker.rs:164]: I'll significantly improve this in a future PR where I'll use a separate set/unset list.

File: lib/gridstore/src/tracker.rs [hunk 3]
```diff
@@ -124,20 +124,69 @@ impl PointerUpdates {
...
+        self.history.is_empty()
+    }
+
+    /// Remove persisted entries from current so that we don't persist them again
+    fn update_from_persisted(&mut self, persisted: &Self) {
+        debug_assert_eq!(
+            self.history.iter().copied().collect::<AHashSet<_>>().len(),
+            self.history.len(),
+            "self must not have duplicate pointers in history",
+        );
```
Comment (timvisee) [inline lib/gridstore/src/tracker.rs:138]: We also do the same debug assertion in the caller of this function. But I vote to keep it for now, until I more properly refactor it later. It does not affect release builds.

File: lib/gridstore/src/tracker.rs [hunk 4]
```diff
@@ -124,20 +124,69 @@ impl PointerUpdates {
...
+
+        let (persist_set, persist_unset) = persisted.to_set_unset();
+
+        // Update current set if necessary
+        match (self.latest(), persist_set) {
+            // Current and persisted set are equal, pop set from current
+            (Some(last), Some(set)) if last == set => {
+                self.history.pop();
+                self.latest_is_set = false;
+            }
```
Comment (coszio) [inline lib/gridstore/src/tracker.rs:148]: Isn't this the same as checking if the entire `PointerUpdates` is the same? In such case we can return early here, since the rest of the code will remove the remaining history.

Comment (timvisee) [reply lib/gridstore/src/tracker.rs:148]: I also thought so.

But in this function I implemented the complete thing on purpose, because I don't want to make any assumptions anymore. Don't want to spend another week on it in the future. :upside_down_face:
